### PR TITLE
Rewrite 1D dilated convolution

### LINKED ISSUE #38638: Dilated convolution pass not working on standard TCN model
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): source
- TensorFlow version (or github SHA if from source): 17dfa4e121c080a547e9cf6443b8fe2ae9ed45ed

**Command used to run the converter or code if you’re using the Python API**
If possible, please share a link to Colab/Jupyter/any notebook.

```
# Copy and paste here the exact command (in Python)
model = load_model(model_path, custom_objects={'TCN': TCN})
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.experimental_new_converter = True
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
```
```
# Copy and paste here the exact command (tf_tfl_translate command-line utility)
tf_tfl_translate  --tf-input-arrays=input_1 --tf-input-shapes=1,784,1 --tf-output-arrays=dense/BiasAdd --print-function-result-mapping -o=$HOME/Desktop/converted.tflite --emit-builtin-tflite-ops ~/smnist.pb
```
**The output from the converter invocation**

```
# Copy and paste the output here. (tf_tfl_translate output)
'main' inputs:
	name: 'input_1' buffer: 0
'main' outputs:
	name: 'dense/BiasAdd' buffer: 271 loc("dense/BiasAdd")
```

**Also, please include a link to the saved model or GraphDef**
[smnist.zip](https://github.com/tensorflow/tensorflow/files/4492741/smnist.zip) contains the model in both .h5 and .pb format, and the converted Flatbuffers.

**Failure details**
I'm converting the Keras-TCN model from https://github.com/philipperemy/keras-tcn to TF Lite.

Method 1:
I convert the Keras model (.h5) to a TF Lite Flatbuffer via the Python API as above. The conversion is successful, inference with the model has the correct accuracy, but I want to get rid of the SpaceToBatchNd and BatchToSpaceNd nodes resulting from the Conv1D ops present in the model. Commit f54bb6f5578b931d79884302768996ba1073f685 claims to do so solving issue #29509. However this does not happen to be so in my converted model, which I attach.

Method 2:
To investigate further, I built the tf_tfl_translate tool from source and invoked it with the command above on a GraphDef (.pb) of the model. Conversions happen to be successful but again the STB and BTS ops are still present in the TF Lite Flatbuffer. The sequence of ops in the GraphDef seems to comply to the sequences specified in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h. 
However, this time the ExpandDims and Squeeze ops are correctly converted to Reshape operations, as specified here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/lite/transforms/optimize_patterns.td, whereas this would not happen with the method above, and although I would like to avoid this being done on my final Flatbuffer.

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes https://github.com/tensorflow/tensorflow/issues/38638. The pattern is obtained from 

```python
layer = tf.keras.layers.Conv1D(8, 5, 1, dilation_rate=2, padding="SAME", use_bias=False)

@tf.function(input_signature=(tf.TensorSpec(shape=(1, 128, 3)),))
def my_func(x):
  return layer(x)
```

Hi @haozha111, can you take a look at this when time allows? Thank you!

### CODE REVIEW

File: tensorflow/compiler/mlir/lite/transforms/dilated_conv.h [hunk 1]
```diff
@@ -297,12 +318,31 @@ ConvertTFDilatedConvOp<Conv2dOpTy>::ExtractDilationsAttrFromBlockShape(
...
+      // NWC -> NHWC
+      dilation_h_factor = 1;
+      dilation_w_factor =
+          stb_bs_attr.getValue({0}).cast<IntegerAttr>().getInt();
+    } else if (expand_axis == 2) {
+      // NHC -> NHWC
+      dilation_h_factor =
+          stb_bs_attr.getValue({0}).cast<IntegerAttr>().getInt();
+      dilation_w_factor = 1;
+    } else {
```
Comment (haozha111) [inline tensorflow/compiler/mlir/lite/transforms/dilated_conv.h:340]: nit, you can reduce those two returns with only one at the end.

**Review by haozha111 — APPROVED** [review]
Thanks a lot for the change! This will be a great improvement to support 1d Dilated conv.

File: tensorflow/compiler/mlir/lite/transforms/dilated_conv.h (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/dilated_conv.h:340]: Will do!

Comment (WindQAQ) [reply tensorflow/compiler/mlir/lite/transforms/dilated_conv.h]: Updated. Please see if this is better.

Comment (haozha111) [reply tensorflow/compiler/mlir/lite/transforms/dilated_conv.h]: Yeap! It looks better.

**Review by haozha111 — APPROVED** [review]

**Comment by haozha111** [comment]
@joker-eph Hi Mehdi, could you lgtm this change since the pull request needs additional owner approval to proceed. Thanks!

**Review by rthadur — APPROVED** [review]
